marlowe/src/Language/Marlowe/Common.hs

```diff
@@ -0,0 +1,845 @@
+{-# LANGUAGE LambdaCase   #-}
+{-# LANGUAGE NamedFieldPuns   #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns -Wno-name-shadowing #-}
+
+{-| = Marlowe: financial contracts on Cardano Computation Layer
+
+Here we present a reference implementation of Marlowe, domain-specific language targeted at
+the execution of financial contracts in the style of Peyton Jones et al
+on Cardano Computation Layer.
+
+The implementation is based on semantics described in paper
+<https://iohk.io/research/papers/#2WHKDRA8 'Marlowe: financial contracts on blockchain'>
+by Simon Thompson and Pablo Lamela Seijas
+
+== Semantics
+
+Semantics is based on <https://github.com/input-output-hk/marlowe/blob/stable/src/Semantics.hs>
+
+Marlowe Contract execution is a chain of transactions,
+where remaining contract and its state is passed through /Data Script/,
+and actions (i.e. /Choices/ and /Oracle Values/) are passed as
+/Redeemer Script/
+
+/Validation Script/ is always the same Marlowe interpreter implementation, available below.
+
+Both /Redeemer Script/ and /Data Script/ have the same structure:
+@(Input, MarloweData)@
+
+where
+
+* 'Input' contains contract actions (i.e. /Pay/, /Redeem/), /Choices/ and /Oracle Values/,
+* 'MarloweData' contains remaining 'Contract' and its 'State'
+* 'State' is a set of 'Commit's plus a set of made 'Choice's
+
+To spend 'TxOut' secured by Marlowe /Validator Script/, a user must provide /Redeemer Script/
+that is a tuple of an 'Input' and expected output of Marlowe 'Contract' interpretation for
+the given 'Input', i.e. 'Contract' and 'State'.
+
+To ensure that user provides valid remainig 'Contract' and 'State'
+Marlowe /Validator Script/ compares evaluated contract and state with provided by user,
+and rejects a transaction if those don't match.
+
+To ensure that remaining contract's /Data Script/ has the same 'Contract' and 'State'
+as was passed with /Redeemer Script/, we check that /Data Script/ hash is
+the same as /Redeemer Script/.
+That's why those are of the same structure @(Input, MarloweData)@.
+
+== Example
+
+Consider simple payment contract, where Alice commits to pay 100 Ada to Bob before timeout1.
+She can get back committed money if Bob didn't ask for payment before timeout2.
+
+> let Alice = PubKey 1
+> let Bob   = PubKey 2
+> let timeout1 = 23
+> let timeout2 = 50
+> let contract = CommitCash (IdentCC 1) Alice (Value 100) timeout1 timeout2
+>             (Pay (IdentPay 1) Alice Bob (Committed (IdentCC 1)) timeout2 Null)
+>             Null
+
+Alice commits:
+
+> let input = Input (Commit (IdentCC 1) (txHash `signedBy` Alice)) [] []
+
+Bob demands payment:
+
+> let input = Input (Payment (IdentPay 1) (txHash `signedBy` Bob)) [] []
+
+Or, in case Bob didn't demand payment before timeout2, Alice can require a redeem of her commit:
+
+> let input = Input (Redeem (IdentCC 1) (txHash `signedBy` Alice)) [] []
```

Comment:
If you wanted you could actually make these doctests. Maybe overkill for now.

Reply:
Excellent idea, but I feel that's an overkill too.